webapp/migrations/Version20251024122021.php

```diff
@@ -0,0 +1,141 @@
+
+        $this->addSql('ALTER TABLE language CHANGE langid_int langid INT UNSIGNED AUTO_INCREMENT NOT NULL COMMENT \'Language ID\'');
+        $this->addSql('ALTER TABLE contestlanguage CHANGE langid_int langid INT UNSIGNED NOT NULL COMMENT \'Language ID\'');
+        $this->addSql('ALTER TABLE problemlanguage CHANGE langid_int langid INT UNSIGNED NOT NULL COMMENT \'Language ID\'');
+        $this->addSql('ALTER TABLE submission CHANGE langid_int langid INT UNSIGNED DEFAULT NULL COMMENT \'Language ID\'');
+        $this->addSql('ALTER TABLE version CHANGE langid_int langid INT UNSIGNED DEFAULT NULL COMMENT \'Language ID\'');
+
+        $this->addSql('ALTER TABLE language ADD PRIMARY KEY (langid), DROP INDEX langid_int');
+        $this->addSql('ALTER TABLE contestlanguage ADD PRIMARY KEY (cid, langid)');
+        $this->addSql('ALTER TABLE problemlanguage ADD PRIMARY KEY (probid, langid)');
+
+        $this->addSql('ALTER TABLE submission ADD INDEX langid (langid), ADD CONSTRAINT submission_ibfk_4 FOREIGN KEY (langid) REFERENCES language (langid) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE problemlanguage ADD INDEX IDX_46B150BB2271845 (langid), ADD CONSTRAINT FK_46B150BB2271845 FOREIGN KEY (langid) REFERENCES language (langid) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE version ADD INDEX IDX_BF1CD3C32271845 (langid), ADD CONSTRAINT FK_BF1CD3C32271845 FOREIGN KEY (langid) REFERENCES language (langid) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE contestlanguage ADD INDEX IDX_ADCB43232271845 (langid), ADD CONSTRAINT FK_ADCB43232271845 FOREIGN KEY (langid) REFERENCES language (langid) ON DELETE CASCADE');
+    }
+
+    public function down(Schema $schema): void
+    {
+        // When migrating back, we do the logic in reverse but we have custom logic for setting
+        // the string langid. We set it to the external ID except for a set of 10 known languages
+        // where the external ID differs from our previously used langid.
+
+        $this->addSql('ALTER TABLE submission DROP FOREIGN KEY submission_ibfk_4');
+        $this->addSql('ALTER TABLE problemlanguage DROP FOREIGN KEY FK_46B150BB2271845');
+        $this->addSql('ALTER TABLE version DROP FOREIGN KEY FK_BF1CD3C32271845');
+        $this->addSql('ALTER TABLE contestlanguage DROP FOREIGN KEY FK_ADCB43232271845');
+
+        $this->addSql('ALTER TABLE language ADD langid_str VARCHAR(32) NOT NULL AFTER langid');
+        $this->addSql('ALTER TABLE contestlanguage ADD langid_str VARCHAR(32) NOT NULL AFTER langid');
+        $this->addSql('ALTER TABLE problemlanguage ADD langid_str VARCHAR(32) NOT NULL AFTER langid');
+        $this->addSql('ALTER TABLE submission ADD langid_str VARCHAR(32) DEFAULT NULL AFTER langid');
+        $this->addSql('ALTER TABLE version ADD langid_str VARCHAR(32) DEFAULT NULL AFTER langid');
+
+        $this->addSql("UPDATE language SET langid_str = CASE externalid
+            WHEN 'ada' THEN 'adb'
```

Comment:
huh, ada -> adb? Is that what you want?

Reply:
Yes, we used `adb` as langid for ADA files, but the CLICS spec uses `ada`.